### PR TITLE
1252

### LINKED ISSUE #1154: Do not works default Vividus` timeout in step which switches to another page.
**Description:** 
When I switch to window with title that is equal to `...` step lasts infinitely while, and looks like default timeout have not occur.

**Main page:** https://amway.uat.hyperwallet.com/hw2web/landing.xhtml?faces-redirect=true&refreshme=true

**STR:**
1. Make local run of such scenario:
Scenario: 1
Given I am on a page with the URL 'https://amway.uat.hyperwallet.com/merchantLogin.jsp'
When I switch to window with title that is equal to `pay-hyperwallet - Merchant Login`
When I wait until element located `By.xpath(//input[@name="LOGIN"])` appears
When I open URL `https://amway.uat.hyperwallet.com/hw2web/landing.xhtml?faces-redirect=true&refreshme=true` in new window
When I switch to window with title that is equal to `Amway Payment Portal - Welcome`
When I enter `password` in field located `By.xpath(//input[@id="username"])`
When I enter `NePassword` in field located `By.xpath(//input[@id="j_password"])`
When I take a screenshot

Expected result: Scenario done without errors/fails. Login and password were entered into appropriate fields.
Actual result: "Switch" step could not be completed.

Capture1
Capture2

[test.txt](https://github.com/vividus-framework/vividus/files/5568601/test.txt)

### LINKED ISSUE #1236: Adding new step with "Stop loading this page" button functionality
I would like to have new step in Vividus that make possible stop page loading (as it is when you click on the cross in your browser).
Cross

### PR DESCRIPTION
**Pull request by ikalinin1**
Fixes #1154 
Closes #1236

### CODE REVIEW

File: vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/PageSteps.java [hunk 1]
```diff
@@ -327,6 +331,25 @@ public void assertPageTitle(StringComparisonRule comparisonRule, String text)
...
+     * <code>window.stop();</code>
+     */
+    @When("I stop page loading")
+    public void stopPageLoading()
+    {
+        try
+        {
+            Map<String, String> pageStates = javascriptActions.stopPageLoading();
+            LOGGER.info("Page ready state before stop: {}, after stop:{}", pageStates.get("before"),
+                    pageStates.get("after"));
```
Comment (uarlouski) [inline vividus-plugin-web-app/src/main/java/org/vividus/bdd/steps/ui/web/PageSteps.java:345]: use lazy logging

File: vividus-plugin-web-app/src/test/java/org/vividus/bdd/steps/ui/web/PageStepsTests.java [hunk 2]
```diff
@@ -330,8 +336,13 @@ void testRefreshPage()
...
     void testOpenPageUrlInNewWindow()
     {
+        InOrder inOrder = Mockito.inOrder(setContextSteps, navigateActions, javascriptActions, uiContext);
         pageSteps.openPageUrlInNewWindow(URL);
-        verify(javascriptActions).openPageUrlInNewWindow(URL);
+        inOrder.verify(javascriptActions).openNewWindow();
+        inOrder.verify(setContextSteps).switchingToWindow();
+        inOrder.verify(uiContext).reset();
+        inOrder.verify(navigateActions).loadPage(URL);
+        verifyNoMoreInteractions(setContextSteps, navigateActions, javascriptActions, uiContext);
```
Comment (uarlouski) [inline vividus-plugin-web-app/src/test/java/org/vividus/bdd/steps/ui/web/PageStepsTests.java:345]: inOrder.verifyNoMoreInteractions should work

File: docs/modules/plugins/pages/plugin-web-app.adoc [hunk 3]
```diff
@@ -219,6 +219,41 @@ include::partial$selenium-properties.adoc[]
 
 == Steps
 
+=== Open URL in a new tab
+
+Loads given URL in a new table
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-web-app.adoc:224]: `table` -> `tab`

File: docs/modules/plugins/pages/plugin-web-app.adoc [hunk 4]
```diff
@@ -219,6 +219,41 @@ include::partial$selenium-properties.adoc[]
...
+
+[source,gherkin]
+----
+When I open URL `$URL` in new window
+----
+* `$URL` - An URL to navigate to
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-web-app.adoc:230]: `The URL to open`

File: docs/modules/plugins/pages/plugin-web-app.adoc [hunk 5]
```diff
@@ -219,6 +219,41 @@ include::partial$selenium-properties.adoc[]
...
+
+.OpenDocsInANewTab.story
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-web-app.adoc:232]: `.Open docs in a new tab.story`

File: docs/modules/plugins/pages/plugin-web-app.adoc [hunk 6]
```diff
@@ -219,6 +219,41 @@ include::partial$selenium-properties.adoc[]
...
+When I open URL `https://docs.vividus.dev` in new window
+----
+
+=== Stop page loading
+
+Stops page loading
+
+[source,gherkin]
+----
+When I step page loading
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-web-app.adoc:244]: `step` -> `stop`

File: docs/modules/plugins/pages/plugin-web-app.adoc [hunk 7]
```diff
@@ -219,6 +219,41 @@ include::partial$selenium-properties.adoc[]
...
+----
+
+[TIP]
+Could be useful in combination with Selenium's https://www.selenium.dev/documentation/en/webdriver/page_loading_strategy/[page load strategy]
+
+.OpenDocsInANewTab.story
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-web-app.adoc:250]: `Stop page loading.story`

File: docs/modules/plugins/pages/plugin-web-app.adoc [hunk 8]
```diff
@@ -219,6 +219,41 @@ include::partial$selenium-properties.adoc[]
...
+[source,gherkin]
+----
+When I open URL `https://delayed.vividus.dev` in new window
+When I step page loading
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-web-app.adoc:254]: `step` -> `stop`

File: vividus-plugin-web-app/src/main/java/org/vividus/ui/web/action/WebJavascriptActions.java [hunk 9]
```diff
@@ -161,6 +160,16 @@ public String getPageText()
...
     }
 
+    public Map<String, String> stopPageLoading()
+    {
+        return executeScript(
+            "let result = {};"
+          + "result.before = document.readyState;"
+          + "window.stop();"
+          + "result.after = document.readyState;"
+          + "return result;");
```
Comment (valfirst) [inline vividus-plugin-web-app/src/main/java/org/vividus/ui/web/action/WebJavascriptActions.java:170]:
```js
let before = document.readyState;
window.stop();
return {before: before, after: document.readyState};
```
?

File: vividus-plugin-web-app/src/test/java/org/vividus/bdd/steps/ui/web/PageStepsTests.java [hunk 10]
```diff
@@ -330,8 +336,13 @@ void testRefreshPage()
     @Test
     void testOpenPageUrlInNewWindow()
     {
+        InOrder inOrder = Mockito.inOrder(setContextSteps, navigateActions, javascriptActions, uiContext);
```
Comment (valfirst) [inline vividus-plugin-web-app/src/test/java/org/vividus/bdd/steps/ui/web/PageStepsTests.java:339]: `inOrder` -> `ordered` (PMD doesn't like variables prefixed with `in` and `out`)

File: vividus-plugin-web-app/src/test/java/org/vividus/ui/web/action/WebJavascriptActionsTests.java [hunk 11]
```diff
@@ -158,11 +158,10 @@ void testScrollToStartOf()
     }
 
     @Test
-    void testOpenPageUrlInNewWindow()
+    void testOpenNewTable()
```
Comment (valfirst) [inline vividus-plugin-web-app/src/test/java/org/vividus/ui/web/action/WebJavascriptActionsTests.java:161]: `Table` -> `Tab`

File: vividus-tests/src/main/resources/story/integration/PageStepsTests.story [hunk 12]
```diff
@@ -15,3 +16,14 @@ When I refresh the page
 When I click on element located `By.id(anchor)`
 When I wait until scroll is finished
 Then the page is scrolled to an element with the attribute 'id'='toClick'
+
+
+Scenario: Verify step: When I open URL `$URL` in new window; Verify step: When I stop page loading
+Meta:
+    @requirementId 1252
```
Comment (valfirst) [inline vividus-tests/src/main/resources/story/integration/PageStepsTests.story:23]: `@requirementId 1154; 1252` ?

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-tests/src/main/resources/story/integration/PageStepsTests.story [hunk 13]
```diff
@@ -15,3 +16,14 @@ When I refresh the page
 When I click on element located `By.id(anchor)`
 When I wait until scroll is finished
 Then the page is scrolled to an element with the attribute 'id'='toClick'
+
+
+Scenario: Verify step: When I open URL `$URL` in new window; Verify step: When I stop page loading
+Meta:
+    @requirementId 1154; 1252
```
Comment (valfirst) [inline vividus-tests/src/main/resources/story/integration/PageStepsTests.story:23]: `1252` -> `1236`

**Review by valfirst — CHANGES_REQUESTED** [review]

**Review by valfirst — APPROVED** [review]

File: vividus-plugin-web-app/src/test/java/org/vividus/bdd/steps/ui/web/PageStepsTests.java [hunk 14]
```diff
@@ -464,4 +475,27 @@ void testPageTitleContainsText()
...
+    }
+
+    @Test
+    void shouldRecordFailedAssertionInCaseOfJavascriptException()
+    {
+        JavascriptException javascriptException = new JavascriptException("I can't take this any moooore");
+        when(javascriptActions.stopPageLoading()).thenThrow(javascriptException);
+        pageSteps.stopPageLoading();
+        softAssert.recordFailedAssertion("Unable to stop page loading", javascriptException);
+        verifyNoMoreInteractions(javascriptActions);
```
Comment (uarlouski) [inline vividus-plugin-web-app/src/test/java/org/vividus/bdd/steps/ui/web/PageStepsTests.java:499]: logger verification is missing :)

**Review by uarlouski — APPROVED** [review]

File: vividus-plugin-web-app/src/test/java/org/vividus/bdd/steps/ui/web/PageStepsTests.java (hunk 14 above)
Comment (ikalinin1) [reply vividus-plugin-web-app/src/test/java/org/vividus/bdd/steps/ui/web/PageStepsTests.java:499]: Wow :)

**Pull request merged**